go/proxyd/daisychain.go

```diff
@@ -0,0 +1,485 @@
+package proxyd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/gorilla/mux"
+	"github.com/gorilla/websocket"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/rs/cors"
+)
+
+var (
+	// Represents the chain id for the Optimism L2 mainnet
+	MainnetChainId = big.NewInt(10)
+	// Represents the chain id for the Optimism L2 public testnet
+	KovanChainId = big.NewInt(69)
+)
+
+// maxEpoch represents the maximum epoch
+// supported by the daisychain. If the RequestOptions
+// is sent with an epoch larger than this, it is an
+// error
+const maxEpoch = uint(6)
+
+// DaisyChainServer is the main struct representing
+// the daisy chain server. It has a reference to
+// each possible backend
+type DaisyChainServer struct {
+	rpcServer          *http.Server
+	maxBodySize        int64
+	authenticatedPaths map[string]string
+	epoch1             *Backend
+	epoch2             *Backend
+	epoch3             *Backend
+	epoch4             *Backend
+	epoch5             *Backend
+	epoch6             *Backend
+	bedrockCutoffBlock *big.Int
+	l2ChainID          *big.Int
+	upgrader           *websocket.Upgrader
+}
+
+// RequestOptions represents the optional
+// additional RPC parameter than can be sent
+// by a user to the daisychain server. The epoch
+// represents the backend that the request should be
+// sent to.
+type RequestOptions struct {
+	Epoch uint `json:"epoch"`
```

Comment:
nit: Given that `epoch` isn't an uncommon term in Ethereum-land (with the beacon chain), perhaps we should rename this to `op_epoch`, for forwards-compatibility.